File: java/src/main/java/io/github/muthuishere/toolnexus/Tools.java
package io.github.muthuishere.toolnexus;

import io.github.muthuishere.toolnexus.annotations.Param;
import io.github.muthuishere.toolnexus.annotations.ToolMethod;

import java.lang.reflect.Method;
import java.lang.reflect.Parameter;
import java.lang.reflect.Type;
import java.util.ArrayList;
import java.util.Collection;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * Reflection-based native tool collection. Scans an object for
 * {@link ToolMethod @ToolMethod}-annotated methods and turns each into a uniform
 * {@link Tool} (source {@code "native"}), inferring the JSON input schema from the
 * method's parameter types.
 *
 * <p>The Spring-AI {@code @Tool} feel, vendor-neutral.
 */
public final class Tools {
    private Tools() {}

    /** Reflect over all {@code @ToolMethod}-annotated methods on the object. */
    public static List<Tool> fromObject(Object target) {
        List<Tool> out = new ArrayList<>();
        Class<?> cls = target.getClass();
        for (Method method : cls.getMethods()) {
            ToolMethod ann = method.getAnnotation(ToolMethod.class);
            if (ann == null) continue;
            out.add(buildTool(target, method, ann));
        }
        return out;
    }

    private static Tool buildTool(Object target, Method method, ToolMethod ann) {
        String name = ann.name().isEmpty() ? method.getName() : ann.name();
        String description = ann.description();

        Parameter[] params = method.getParameters();
        Map<String, Object> properties = new LinkedHashMap<>();
        List<String> required = new ArrayList<>();
        List<String> paramNames = new ArrayList<>();

        for (Parameter p : params) {
            if (p.getType().equals(ToolContext.class)) {
                paramNames.add("$ctx");
                continue;
            }
            Param pann = p.getAnnotation(Param.class);
            String pname = (pann != null && !pann.name().isEmpty()) ? pann.name() : p.getName();
            paramNames.add(pname);

            Map<String, Object> prop = jsonTypeOf(p.getType(), p.getParameterizedType());
            if (pann != null && !pann.description().isEmpty()) {
                prop.put("description", pann.description());
            }
            properties.put(pname, prop);

            boolean req = pann == null || pann.required();
            if (req) required.add(pname);
        }

        Map<String, Object> schema = new LinkedHashMap<>();
        schema.put("type", "object");
        schema.put("properties", properties);
        if (!required.isEmpty()) schema.put("required", required);
        schema.put("additionalProperties", false);

        method.setAccessible(true);

        return NativeTool.of(name, description, schema, (args, ctx) -> {
            Map<String, Object> a = args == null ? new LinkedHashMap<>() : args;
            Object[] callArgs = new Object[params.length];
            for (int i = 0; i < params.length; i++) {
                String pn = paramNames.get(i);
                if ("$ctx".equals(pn)) {
                    callArgs[i] = ctx;
                } else {
                    callArgs[i] = coerce(a.get(pn), params[i].getType());
                }
            }
            try {
                return method.invoke(target, callArgs);
            } catch (java.lang.reflect.InvocationTargetException e) {
                Throwable cause = e.getCause() != null ? e.getCause() : e;
                if (cause instanceof RuntimeException) throw (RuntimeException) cause;
                throw new RuntimeException(cause.getMessage(), cause);
            } catch (IllegalAccessException e) {
                throw new RuntimeException(e.getMessage(), e);
            }
        });
    }

    /** Map a Java type to a JSON-Schema property fragment. */
    static Map<String, Object> jsonTypeOf(Class<?> type, Type generic) {
        Map<String, Object> m = new LinkedHashMap<>();
        if (type.equals(String.class) || type.equals(char.class) || type.equals(Character.class)) {
            m.put("type", "string");
        } else if (type.equals(boolean.class) || type.equals(Boolean.class)) {
            m.put("type", "boolean");
        } else if (type.equals(int.class) || type.equals(Integer.class)
                || type.equals(long.class) || type.equals(Long.class)
                || type.equals(double.class) || type.equals(Double.class)
                || type.equals(float.class) || type.equals(Float.class)
                || type.equals(short.class) || type.equals(Short.class)
                || Number.class.isAssignableFrom(type)) {
            m.put("type", "number");
        } else if (Collection.class.isAssignableFrom(type) || type.isArray()) {
            m.put("type", "array");
        } else {
            // Map or POJO -> object
            m.put("type", "object");
        }
        return m;
    }

    /** Best-effort coercion of a JSON-decoded value into the method parameter type. */
    @SuppressWarnings("unchecked")
    static Object coerce(Object value, Class<?> type) {
        if (value == null) {
            if (type.isPrimitive()) {
                if (type.equals(boolean.class)) return false;
                if (type.equals(int.class)) return 0;
                if (type.equals(long.class)) return 0L;
                if (type.equals(double.class)) return 0d;
                if (type.equals(float.class)) return 0f;
                if (type.equals(short.class)) return (short) 0;
                if (type.equals(char.class)) return '\0';
            }
            return null;
        }
        if (type.isInstance(value)) return value;

        if (type.equals(String.class)) return String.valueOf(value);

        if (value instanceof Number) {
            Number n = (Number) value;
            if (type.equals(int.class) || type.equals(Integer.class)) return n.intValue();
            if (type.equals(long.class) || type.equals(Long.class)) return n.longValue();
            if (type.equals(double.class) || type.equals(Double.class)) return n.doubleValue();
            if (type.equals(float.class) || type.equals(Float.class)) return n.floatValue();
            if (type.equals(short.class) || type.equals(Short.class)) return n.shortValue();
        }
        if (value instanceof String) {
            String s = (String) value;
            try {
                if (type.equals(int.class) || type.equals(Integer.class)) return Integer.parseInt(s);
                if (type.equals(long.class) || type.equals(Long.class)) return Long.parseLong(s);
                if (type.equals(double.class) || type.equals(Double.class)) return Double.parseDouble(s);
                if (type.equals(float.class) || type.equals(Float.class)) return Float.parseFloat(s);
                if (type.equals(short.class) || type.equals(Short.class)) return Short.parseShort(s);
                if (type.equals(boolean.class) || type.equals(Boolean.class)) return Boolean.parseBoolean(s);
            } catch (NumberFormatException ignored) {
            }
        }
        if (value instanceof Boolean && (type.equals(boolean.class) || type.equals(Boolean.class))) {
            return value;
        }
        return value;
    }
}

File: csharp/src/Toolnexus/Tools.cs
using System.Reflection;
using System.Text.RegularExpressions;

namespace Toolnexus;

/// <summary>
/// Tool helpers: the shared <see cref="Sanitize"/> name rule and reflection-based
/// native tool collection (<see cref="FromObject"/>). Scans an object for
/// <see cref="ToolMethodAttribute"/>-annotated methods and turns each into a uniform
/// <see cref="ITool"/> (source <c>"native"</c>), inferring the JSON input schema from
/// the method's parameter types.
/// </summary>
public static partial class Tools
{
    [GeneratedRegex("[^a-zA-Z0-9_-]")]
    private static partial Regex SanitizeRegex();

    /// <summary>Replace anything outside <c>[a-zA-Z0-9_-]</c> with <c>"_"</c> (same as opencode).</summary>
    public static string Sanitize(string? value)
        => value == null ? "" : SanitizeRegex().Replace(value, "_");

    /// <summary>Reflect over all <see cref="ToolMethodAttribute"/>-annotated methods on the object.</summary>
    public static List<ITool> FromObject(object target)
    {
        var output = new List<ITool>();
        foreach (var method in target.GetType().GetMethods(BindingFlags.Public | BindingFlags.Instance))
        {
            var ann = method.GetCustomAttribute<ToolMethodAttribute>();
            if (ann == null) continue;
            output.Add(BuildTool(target, method, ann));
        }
        return output;
    }

    private static ITool BuildTool(object target, MethodInfo method, ToolMethodAttribute ann)
    {
        var name = string.IsNullOrEmpty(ann.Name) ? method.Name : ann.Name;
        var description = ann.Description;

        var parameters = method.GetParameters();
        var properties = new Dictionary<string, object?>();
        var required = new List<object?>();
        var paramNames = new List<string>();

        foreach (var p in parameters)
        {
            if (p.ParameterType == typeof(ToolContext))
            {
                paramNames.Add("$ctx");
                continue;
            }

            var pann = p.GetCustomAttribute<ParamAttribute>();
            var pname = pann != null && !string.IsNullOrEmpty(pann.Name) ? pann.Name : p.Name ?? "arg";
            paramNames.Add(pname);

            var prop = JsonTypeOf(p.ParameterType);
            if (pann != null && !string.IsNullOrEmpty(pann.Description))
                prop["description"] = pann.Description;
            properties[pname] = prop;

            var req = pann?.Required ?? true;
            if (req) required.Add(pname);
        }

        var schema = new Dictionary<string, object?>
        {
            ["type"] = "object",
            ["properties"] = properties,
        };
        if (required.Count > 0) schema["required"] = required;
        schema["additionalProperties"] = false;

        return NativeTool.OfAsync(name, description, schema, async (args, ctx) =>
        {
            var a = args ?? new Dictionary<string, object?>();
            var callArgs = new object?[parameters.Length];
            for (var i = 0; i < parameters.Length; i++)
            {
                var pn = paramNames[i];
                if (pn == "$ctx")
                    callArgs[i] = ctx;
                else
                    callArgs[i] = Coerce(a.TryGetValue(pn, out var v) ? v : null, parameters[i].ParameterType);
            }

            object? result;
            try
            {
                result = method.Invoke(target, callArgs);
            }
            catch (TargetInvocationException tie)
            {
                throw tie.InnerException ?? tie;
            }

            // Await Task / Task<T> returns.
            if (result is Task task)
            {
                await task.ConfigureAwait(false);
                var taskType = task.GetType();
                if (taskType.IsGenericType)
                    result = taskType.GetProperty("Result")!.GetValue(task);
                else
                    result = null;
            }
            return result;
        });
    }

    /// <summary>Map a CLR type to a JSON-Schema property fragment.</summary>
    internal static Dictionary<string, object?> JsonTypeOf(Type type)
    {
        var t = Nullable.GetUnderlyingType(type) ?? type;
        var m = new Dictionary<string, object?>();
        if (t == typeof(string) || t == typeof(char))
            m["type"] = "string";
        else if (t == typeof(bool))
            m["type"] = "boolean";
        else if (t == typeof(int) || t == typeof(long) || t == typeof(double) || t == typeof(float)
                 || t == typeof(short) || t == typeof(decimal) || t == typeof(byte))
            m["type"] = "number";
        else if (t.IsArray || (typeof(System.Collections.IEnumerable).IsAssignableFrom(t) && t != typeof(string)
                 && !typeof(System.Collections.IDictionary).IsAssignableFrom(t)))
            m["type"] = "array";
        else
            m["type"] = "object";
        return m;
    }

    /// <summary>Best-effort coercion of a JSON-decoded value into the method parameter type.</summary>
    internal static object? Coerce(object? value, Type type)
    {
        var t = Nullable.GetUnderlyingType(type) ?? type;
        if (value == null)
        {
            if (t.IsValueType) return Activator.CreateInstance(t);
            return null;
        }
        if (t.IsInstanceOfType(value)) return value;

        if (t == typeof(string)) return value.ToString();

        try
        {
            if (t == typeof(int)) return Convert.ToInt32(value);
            if (t == typeof(long)) return Convert.ToInt64(value);
            if (t == typeof(double)) return Convert.ToDouble(value);
            if (t == typeof(float)) return Convert.ToSingle(value);
            if (t == typeof(short)) return Convert.ToInt16(value);
            if (t == typeof(decimal)) return Convert.ToDecimal(value);
            if (t == typeof(bool))
            {
                if (value is bool b) return b;
                return bool.TryParse(value.ToString(), out var parsed) && parsed;
            }
        }
        catch
        {
            // fall through to returning the original value
        }
        return value;
    }
}
